asserts/snapasserts/validation_sets.go

```diff
@@ -1031,33 +990,137 @@ func (v *ValidationSets) SnapConstrained(snapRef naming.SnapRef) bool {
 	return v.constraintsForSnap(snapRef) != nil
 }
 
-// CheckPresenceInvalid returns the list of all validation sets that declare
-// presence of the given snap as invalid. PresenceConstraintError is returned if
-// presence of the snap is "optional" or "required".
-// The method assumes that validation sets are not in conflict.
-func (v *ValidationSets) CheckPresenceInvalid(snapRef naming.SnapRef) ([]ValidationSetKey, error) {
-	cstrs := v.constraintsForSnap(snapRef)
+type presence struct {
+	Presence asserts.Presence
+	Revision snap.Revision
+	Sets     ValidationSetKeySlice
+}
+
+// SnapPresence contains information about a snap's allowed presence with
+// respect to a set of validation sets.
+type SnapPresence struct {
+	presence
+	components map[string]presence
+}
+
+// Constrained returns true if the snap is constrained in any way by the
+// validation sets that this SnapPresence is created from.
+// Ultimately, one of these things must be true for a snap to be constrained:
+//   - snap has a presence of either "required" or "invalid"
+//   - the snap's revision is pinned to a specific revision
+//   - either of the above are true for any of the snap's components
+func (s *SnapPresence) Constrained() bool {
+	if s.constrained() {
+		return true
+	}
+
+	for _, comp := range s.components {
+		if comp.constrained() {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *presence) constrained() bool {
+	return p.Presence != asserts.PresenceOptional || !p.Revision.Unset()
+}
+
+// Component returns the presence of the given component of the snap. If this
+// SnapPresence doesn't know about the component, the component will be
+// considered optional and allowed to have any revision.
+func (s *SnapPresence) Component(name string) presence {
+	if s.components == nil {
+		return presence{
+			Presence: asserts.PresenceOptional,
+		}
+	}
+
+	cp, ok := s.components[name]
+	if !ok {
+		return presence{
+			Presence: asserts.PresenceOptional,
+		}
+	}
+	return cp
+}
+
+// RequiredComponents returns a set of all of the components that are required
+// to be installed when this snap is installed.
+func (s *SnapPresence) RequiredComponents() map[string]presence {
```

Comment:
it's strange to have a public method that has a return that uses an unexported type?